### PR TITLE
Add support for MySQL semi-sync cluster

### PR DESCRIPTION
**Pull request by heheh13**

### CODE REVIEW

**Review by spectro30 — DISMISSED** [review]
LGTM

**Comment by tamalsaha** [comment]
Do we support both semi sync and async?

**Comment by heheh13** [comment]
we didn't add support for async separately, our read replica uses async replication

File: apis/kubedb/v1alpha2/mysql_types.go [hunk 1]
```diff
@@ -180,6 +181,25 @@ type MySQLTopology struct {
 	// and it will take reference of  appbinding of the source
 	// +optional
 	ReadReplica *MySQLReadReplicaSpec `json:"readReplica,omitempty"`
+	// +optional
+	SemiSync *SemiSyncSpec `json:"semiSync,omitempty"`
+}
+type ErrantTransactionRecoveryPolicy string
```
Comment (tamalsaha) [inline apis/kubedb/v1alpha2/mysql_types.go:187]: // +kubebuilder:validation:Enum= Clone; PseudoTransaction

File: apis/kubedb/v1alpha2/mysql_types.go [hunk 2]
```diff
@@ -180,6 +181,25 @@ type MySQLTopology struct {
...
+	ErrantTransactionRecoveryPolicyClone             ErrantTransactionRecoveryPolicy = "Clone"
+	ErrantTransactionRecoveryPolicyPseudoTransaction ErrantTransactionRecoveryPolicy = "PseudoTransaction"
+)
+
+type SemiSyncSpec struct {
+	// count of slave to wait for before commit
+	// +kubebuilder:default=1
+	SemiSyncMasterWaitFoSlaveCount *int `json:"semiSyncMasterWaitFoSlaveCount,omitempty"`
+	// +kubebuilder:default=86400
+	SemiSyncMasterTimeout *int `json:"semiSyncMasterTimeout,omitempty"`
```
Comment (tamalsaha) [inline apis/kubedb/v1alpha2/mysql_types.go:199]: `metav1.Duration`

File: apis/kubedb/v1alpha2/mysql_types.go [hunk 3]
```diff
@@ -180,6 +181,25 @@ type MySQLTopology struct {
...
+
+const (
+	ErrantTransactionRecoveryPolicyClone             ErrantTransactionRecoveryPolicy = "Clone"
+	ErrantTransactionRecoveryPolicyPseudoTransaction ErrantTransactionRecoveryPolicy = "PseudoTransaction"
+)
+
+type SemiSyncSpec struct {
+	// count of slave to wait for before commit
+	// +kubebuilder:default=1
+	SemiSyncMasterWaitFoSlaveCount *int `json:"semiSyncMasterWaitFoSlaveCount,omitempty"`
```
Comment (tamalsaha) [inline apis/kubedb/v1alpha2/mysql_types.go:197]: `*int` -> `int`

File: apis/kubedb/v1alpha2/mysql_types.go [hunk 4]
```diff
@@ -180,6 +181,25 @@ type MySQLTopology struct {
...
+	// +kubebuilder:default=86400
+	SemiSyncMasterTimeout *int `json:"semiSyncMasterTimeout,omitempty"`
+	// recovery method if the slave has any errant transaction
+	// +kubebuilder:default=PseudoTransaction
```
Comment (tamalsaha) [inline apis/kubedb/v1alpha2/mysql_types.go:201]: "PseudoTransaction"

File: apis/kubedb/v1alpha2/mysql_types.go [hunk 5]
```diff
@@ -180,6 +181,29 @@ type MySQLTopology struct {
...
+const (
+	ErrantTransactionRecoveryPolicyClone             ErrantTransactionRecoveryPolicy = "Clone"
+	ErrantTransactionRecoveryPolicyPseudoTransaction ErrantTransactionRecoveryPolicy = "PseudoTransaction"
+)
+
+type SemiSyncSpec struct {
+	// count of slave to wait for before commit
+	// +kubebuilder:default=1
+	//+kubebuilder:valida+kubebuilder:validation:Minimum=1
+	SourceWaitFoReplicaCount int `json:"SourceWaitFoReplicaCount,omitempty"`
```
Comment (tamalsaha) [inline apis/kubedb/v1alpha2/mysql_types.go:201]: `json:"SourceWaitFoReplicaCount

should be small `s` `sourceWaitFoReplicaCount`

File: apis/kubedb/v1alpha2/mysql_types.go [hunk 6]
```diff
@@ -180,6 +181,29 @@ type MySQLTopology struct {
...
+	// +kubebuilder:default=86400
+	SourceTimeoutSecond int `json:"SourceTimeoutSecond,omitempty"`
```
Comment (tamalsaha) [inline apis/kubedb/v1alpha2/mysql_types.go:203]: `json:"sourceTimeoutSecond

**Pull request merged**